In [1]:
import math
def f(x,t):
  if (t==0):
    return x*math.log2(x)
  elif (t==1):
    return math.factorial(int(x))
  else:
    return int(x)**int(x)

def find(t,v):
  if (t==0):
    r = v
  else:
    r = 100
  l = 1
  
  while True:
    m = int((l+r)/2)
    if ((f(m+1,t) > v) and (f(m,t) <= v) ) or (l == m) or  (r == m):
      return int(m)   
    elif f(m,t) < v:
      l = m
    else: 
      r = m

$\bf{ Time\ Complexity\ \lg^2(n) }$

In [2]:
print("Second: 2^{",round(math.sqrt(10**9)),"}")
print("Minute: 2^{",round(math.sqrt(60*10**9)),"}")
print("Hour: 2^{",round(math.sqrt(60*60*10**9)),"}")
print("Day: 2^{",round(math.sqrt(24*60*60*10**9)),"}")
print("Year: 2^{",round(math.sqrt(365*24*60*60*10**9)),"}")
print("Century: 2^{",round(math.sqrt(100*365*24*60*60*10**9)),"}")
print("Mileniun: 2^{",round(math.sqrt(1000*100*365*24*60*60*10**9)),"}")

Second: 2^{ 31623 }
Minute: 2^{ 244949 }
Hour: 2^{ 1897367 }
Day: 2^{ 9295160 }
Year: 2^{ 177583783 }
Century: 2^{ 1775837830 }
Mileniun: 2^{ 56156922993 }


$\bf{ Time\ Complexity\ \sqrt{n} }$

In [3]:
print("Second :{:.8E}".format(math.pow(10**9,2)))
print("Minute  :{:.8E}".format(math.pow(60*10**9,2)))
print("Hour :{:.8E}".format(math.pow(60*60*10**9,2)))
print("Day :{:.8E}".format(math.pow(24*60*60*10**9,2)))
print("Year :{:.8E}".format(math.pow(365*24*60*60*10**9,2)))
print("Century :{:.8E}".format(math.pow(100*365*24*60*60*10**9,2)))
print("Mileniun  :{:.8E}".format(math.pow(1000*100*365*24*60*60*10**9,2)))

Second :1.00000000E+18
Minute  :3.60000000E+21
Hour :1.29600000E+25
Day :7.46496000E+27
Year :9.94519296E+32
Century :9.94519296E+36
Mileniun  :9.94519296E+42


$\bf{ Time\ Complexity\ n }$

In [4]:
print("Second: {:.8E}".format(10**9))
print("Minute: {:.8E}".format(60*10**9))
print("Hour: {:.8E}".format(60*60*10**9))
print("Day: {:.8E}".format(24*60*60*10**9))
print("Year: {:.8E}".format(365*24*60*60*10**9))
print("Century: {:.8E}".format(100*365*24*60*60*10**9))
print("Mileniun: {:.8E}".format(1000*100*365*24*60*60*10**9))

Second: 1.00000000E+09
Minute: 6.00000000E+10
Hour: 3.60000000E+12
Day: 8.64000000E+13
Year: 3.15360000E+16
Century: 3.15360000E+18
Mileniun: 3.15360000E+21


$\bf{ Time\ Complexity\ n\lg{n} }$

In [5]:
print("Second: {:.8E}".format(find(0,10**9 )))
print("Minute:  {:.8E}".format(find(0,60*10**9 )))
print("Hour: {:.8E}".format(find(0,60*60*10**9 )))
print("Day: {:.8E}".format(find(0,24*60*60*10**9 )) )
print("Year: {:.8E}".format(find(0,365*24*60*60*10**9 )) )
print("Century: {:.8E}".format(find(0,100*365*24*60*60*10**9 )) )
print("Mileniun: {:.8E}".format(find(0,1000*100*365*24*60*60*10**9 )) )

Second: 3.96200770E+07
Minute:  1.94447045E+09
Hour: 9.85747745E+10
Day: 2.11037274E+12
Year: 6.41136862E+14
Century: 5.66650934E+16
Mileniun: 4.82300896E+19


$\bf{ Time\ Complexity\ n^2}$

In [6]:
print("Second: ",round(math.sqrt(10**9 )))
print("Minute: ",round(math.sqrt(60*10**9 )))
print("Hour: ",round(math.sqrt(60*60*10**9 )))
print("Day: ",round(math.sqrt(24*60*60*10**9 )))
print("Year: ",round(math.sqrt(365*24*60*60*10**9 )))
print("Century: ",round(math.sqrt(100*365*24*60*60*10**9 )))
print("Mileniun: ",round(math.sqrt(1000*100*365*24*60*60*10**9 )))

Second:  31623
Minute:  244949
Hour:  1897367
Day:  9295160
Year:  177583783
Century:  1775837830
Mileniun:  56156922993


$\bf{ Time\ Complexity\ n^3 }$

In [7]:
print("Second: ",round(math.pow(10**9 ,1/3)))
print("Minute:  ",round(math.pow(60*10**9 ,1/3)))
print("Hour: ",round(math.pow(60*60*10**9 ,1/3)))
print("Day: ",round(math.pow(24*60*60*10**9,1/3)))
print("Year: ",round(math.pow(365*24*60*60*10**9 ,1/3)))
print("Century: ",round(math.pow(100*365*24*60*60*10**9 ,1/3)))
print("Mileniun: ",round(math.pow(1000*100*365*24*60*60*10**9 ,1/3)))

Second:  1000
Minute:   3915
Hour:  15326
Day:  44208
Year:  315938
Century:  1466455
Mileniun:  14664554


$\bf{ Time\ Complexity\ n^4 }$

In [8]:
print("Second: ",round(math.pow(10**9 ,1/4)))
print("Minute: ",round(math.pow(60*10**9 ,1/4)))
print("Hour: ",round(math.pow(60*60*10**9 ,1/4)))
print("Day: ",round(math.pow(24*60*60*10**9 ,1/4)))
print("Year: ",round(math.pow(365*24*60*60*10**9 ,1/4)))
print("Century: ",round(math.pow(100*365*24*60*60*10**9 ,1/4)))
print("Mileniun: ",round(math.pow(1000*100*365*24*60*60*10**9 ,1/4)))

Second:  178
Minute:  495
Hour:  1377
Day:  3049
Year:  13326
Century:  42141
Mileniun:  236975


$\bf{ Time\ Complexity\ 2^n }$

In [9]:
print("Second: ",int(math.log2(10**9 )))
print("Minute: ",int(math.log2(60*10**9 )))
print("Hour: ",int(math.log2(60*60*10**9 )))
print("Day: ",int(math.log2(24*60*60*10**9 )))
print("Year: ",int(math.log2(365*24*60*60*10**9 )))
print("Century: ",int(math.log2(100*365*24*60*60*10**9 )))
print("Mileniun: ",int(math.log2(1000*100*365*24*60*60*10**9 )))

Second:  29
Minute:  35
Hour:  41
Day:  46
Year:  54
Century:  61
Mileniun:  71


$\bf{ Time\ Complexity\ n! }$

In [10]:
print("Second: ", int(find(1,10**9 )))
print("Minute:  ",int(find(1,60*10**9 )))
print("Hour: ",int(find(1,60*60*10**9 )))
print("Day: ",int(find(1,24*60*60*10**9 )))
print("Year: ",int(find(1,365*24*60*60*10**9 )))
print("Century: ",int(find(1,100*365*24*60*60*10**9 )))
print("Mileniun: ",int(find(1,1000*100*365*24*60*60*10**9 )))

Second:  12
Minute:   13
Hour:  15
Day:  16
Year:  18
Century:  20
Mileniun:  22


$\bf{ Time\ Complexity\ n^n }$

In [11]:
print("Second: ", find(2,10**9 ))
print("Minute: ",find(2,60*10**9 ))
print("Hour: ",find(2,(60*60)*10**9 ))
print("Day: ",find(2,(24*60*60)*10**9 ))
print("Year: ",find(2,(365*24*60*60)*10**9 ))
print("Century: ",find(2,(100*365*24*60*60)*10**9 ))
print("Mileniun: ",find(2,(1000*100*365*24*60*60)*10**9 ))

Second:  9
Minute:  10
Hour:  11
Day:  12
Year:  14
Century:  15
Mileniun:  17


# 3. (1 nanosecond by instruction)

Asuuming that you can execute a 1000 million instructions $10^9$ every second, i.e. each instrcution takes 1  nanosecond, 1 $\eta s  = 10^{-9} s$. If you have algorithms take $\bf{ Time\ Complexity}$ steps to solve a problem of size $n$ as specified in column one. What is the size of the largest problem that you can solve in the time specified in row one?
$$
\begin{array}{|c|c|c|c|c|c|c|c|} 
 \hline
 \bf{ Time\ Complexity} & Second  & Minute  & Hour & Day & Year & Century & Mileniun  \\ 
  \hline\\ 
\lg(n) &   2^{10^{9} }  &  2^{6 \times 10^{10} } & 2^{3.6 \times 10^{12}} &  2^{8.64 \times 10^{13}}& 2^{3.1536 \times 10^{16}} & 2^{3.1536 \times 10^{18}} & 2^{3.1536 \times 10^{21}} \\ 
   \hline\\ 
\lg^2(n) &   2^{10^{(9/2)}}  & 2^{10^5 \times\sqrt6}  & 2^{10^5 \times\sqrt6} & 2^{ 10^{13/2}\times\sqrt{8.64}  } & 2^{ 10^8\times\sqrt{3.1536} } & 2^{ 10^9\times\sqrt{3.1536} } &  2^{ 10^{10}\times\sqrt{31.536} } \\ 
   \hline\\ 
\sqrt{n} &  10^{18}  & 3.6 \times10^{21}  & 1.296 \times10^{25} & 7.15716 \times10^{27}& 9.94519296 \times10^{32} & 9.94519296  \times10^{36} & 9.94519296 \times10^{42}\\ 
 \hline
 n & 10^{9}    & 6 \times 10^{10}   & 3.6 \times 10^{12} & 8.64 \times 10^{13} & 3.1536 \times 10^{16} & 3.1536 \times 10^{18} &  3.1536 \times 10^{21}\\ 
 \hline
 n\lg(n) &  39620077  & 1944470450  &  98574774544  & 2110372739876&  641136862352841 & 56665093434063120 &  48230089612197773312 \\ 
 \hline
 n^2 &  31623  & 244949  & 1897367 & 9295160 & 177583783 & 1775837830 &  56156922993 \\ 
 \hline
  n^3 &  215 &  843 & 3302 &  9524  &  68067 & 315938 & 3159382 \\ 
 \hline
  n^4 &  178  &  495  & 1377 & 3049 & 13326 & 42141  & 236975 \\ 
 \hline
  2^n &  29  &  35 & 41 & 46 & 54 & 61 & 71 \\ 
 \hline
 n! &  12  & 13 &  15 & 16 & 18 & 20  &  22\\ 
 \hline
 n^n &  9  &  10 & 11 & 12 & 14 & 15 & 17 \\ 
 \hline
\end{array}
$$